File: backend/src/platform/schemas/billing.py
from datetime import date, datetime

from pydantic import BaseModel, ConfigDict, Field, computed_field

from src.platform.core.config import settings
from src.platform.schemas.types import ConstrainedEmail


class CheckoutIn(BaseModel):
    plan_price_id: int


class SwitchPlanIn(BaseModel):
    plan_price_id: int


class StartTrialIn(BaseModel):
    plan_price_id: int


class UpdateBillingEmailIn(BaseModel):
    billing_email: ConstrainedEmail


class PlanPriceOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    plan_id: int
    amount: int
    currency: str
    interval: str
    interval_count: int
    is_active: bool

    @computed_field
    @property
    def trial_period_days(self) -> int | None:
        return settings.billing_trial_period_days or None

    created_at: datetime
    updated_at: datetime


class PlanSettingOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    key: str
    value: int | None


class PlanOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    name: str
    description: str | None
    is_active: bool
    prices: list[PlanPriceOut]
    plan_settings: list[PlanSettingOut]
    created_at: datetime
    updated_at: datetime


class SubscriptionOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    organization_id: int
    plan_price_id: int | None
    status: str
    current_period_start: datetime | None
    current_period_end: datetime | None
    cancel_at_period_end: bool
    canceled_at: datetime | None
    cancel_at: datetime | None
    trial_end: datetime | None
    plan_price: PlanPriceOut | None
    billing_email: str | None = None
    has_payment_method: bool = False
    trial_used: bool = False
    features: list[str] = Field(default_factory=list)
    plan_settings: list[PlanSettingOut] = Field(default_factory=list)
    created_at: datetime
    updated_at: datetime


class UsageItemOut(BaseModel):
    metric: str
    count: int
    limit: int | None


class UsageOut(BaseModel):
    period_start: date
    usage: list[UsageItemOut]


class CheckoutOut(BaseModel):
    checkout_url: str


class CustomerPortalOut(BaseModel):
    portal_url: str

File: frontend/src/platform/types/billing.ts
export interface PlanPriceOut {
  id: number
  plan_id: number
  amount: number
  currency: string
  interval: 'month' | 'year'
  interval_count: number
  trial_period_days: number | null
  is_active: boolean
  created_at: string
  updated_at: string
}

export interface PlanSettingOut {
  key: string
  value: number | null
}

export interface PlanOut {
  id: number
  name: string
  description: string | null
  is_active: boolean
  prices: PlanPriceOut[]
  plan_settings: PlanSettingOut[]
  created_at: string
  updated_at: string
}

export interface SubscriptionOut {
  id: number
  organization_id: number
  plan_price_id: number | null
  status: 'incomplete' | 'active' | 'trialing' | 'past_due' | 'canceled' | 'paused'
  current_period_start: string | null
  current_period_end: string | null
  cancel_at_period_end: boolean
  canceled_at: string | null
  cancel_at: string | null
  trial_end: string | null
  plan_price: PlanPriceOut | null
  billing_email: string | null
  has_payment_method: boolean
  trial_used: boolean
  features: string[]
  plan_settings: PlanSettingOut[]
  created_at: string
  updated_at: string
}

export interface UsageItemOut {
  metric: string
  count: number
  limit: number | null
}

export interface UsageOut {
  period_start: string
  usage: UsageItemOut[]
}

export interface CheckoutOut {
  checkout_url: string
}

export interface CustomerPortalOut {
  portal_url: string
}
